httpd/http_server.c

```diff
@@ -0,0 +1,346 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "daemon/common.h"
+#include "http_server.h"
+#include "h2o.h"
+
+#include "h2o_utils.h"
+
+static h2o_globalconf_t config;
+static h2o_context_t ctx;
+static h2o_accept_ctx_t accept_ctx;
+
+#define CONTENT_JSON_UTF8 H2O_STRLIT("application/json; charset=utf-8")
+#define CONTENT_TEXT_UTF8 H2O_STRLIT("text/plain; charset=utf-8")
+#define NBUF_INITIAL_SIZE_RESP (4096)
+#define API_V1_PREFIX "/api/v1/"
+#define HOST_SELECT_PREFIX "/host/"
+
+#define HTTPD_CONFIG_SECTION "httpd"
+#define HTTPD_ENABLED_DEFAULT false
+
+static void on_accept(uv_stream_t *listener, int status)
+{
+    uv_tcp_t *conn;
+    h2o_socket_t *sock;
+
+    if (status != 0)
+        return;
+
+    conn = h2o_mem_alloc(sizeof(*conn));
+    uv_tcp_init(listener->loop, conn);
+
+    if (uv_accept(listener, (uv_stream_t *)conn) != 0) {
+        uv_close((uv_handle_t *)conn, (uv_close_cb)free);
+        return;
+    }
+
+    sock = h2o_uv_socket_create((uv_stream_t *)conn, (uv_close_cb)free);
+    h2o_accept(&accept_ctx, sock);
+}
+
+static int create_listener(const char *ip, int port)
+{
+    static uv_tcp_t listener;
+    struct sockaddr_in addr;
+    struct sockaddr_in6 addr6;
+    int r, ip6 = 0;
+
+    uv_tcp_init(ctx.loop, &listener);
+    if (uv_ip4_addr(ip, port, &addr)) {
+        ip6 = 1;
+        if (uv_ip6_addr(ip, port, &addr6)) {
+            error_report("Could not parse the ip address");
+            goto Error;
+        }
+    }
+
+    if ((r = uv_tcp_bind(&listener, (ip6 ? (struct sockaddr *)&addr6 : (struct sockaddr *)&addr), 0)) != 0) {
+        fprintf(stderr, "uv_tcp_bind:%s\n", uv_strerror(r));
+        goto Error;
+    }
+    if ((r = uv_listen((uv_stream_t *)&listener, 128, on_accept)) != 0) {
+        fprintf(stderr, "uv_listen:%s\n", uv_strerror(r));
+        goto Error;
+    }
+
+    return 0;
+Error:
+    uv_close((uv_handle_t *)&listener, NULL);
+    return r;
+}
+
+static int ssl_init()
+{
+    if (!config_get_boolean(HTTPD_CONFIG_SECTION, "ssl", false))
+        return 0;
+
+    char default_fn[FILENAME_MAX + 1];
+
+    snprintfz(default_fn,  FILENAME_MAX, "%s/ssl/key.pem",  netdata_configured_user_config_dir);
+    const char *key_fn  = config_get(HTTPD_CONFIG_SECTION, "ssl key", default_fn);
+
+    snprintfz(default_fn, FILENAME_MAX, "%s/ssl/cert.pem", netdata_configured_user_config_dir);
+    const char *cert_fn = config_get(HTTPD_CONFIG_SECTION, "ssl certificate",  default_fn);
+
+#if OPENSSL_VERSION_NUMBER < OPENSSL_VERSION_110
+    accept_ctx.ssl_ctx = SSL_CTX_new(SSLv23_server_method());
+#else
+    accept_ctx.ssl_ctx = SSL_CTX_new(TLS_server_method());
+#endif
+
+    SSL_CTX_set_options(accept_ctx.ssl_ctx, SSL_OP_NO_SSLv2);
```

Comment:
Like the functions we call above to define the server method, we are also using different [functions](https://github.com/netdata/netdata/blob/0275a0d468d8b41d4b1d84955b99a77c9eed8759/libnetdata/socket/security.c#L90-L107) to set options for context, this can be the source of issues I reported to you.

Reply:
Please, take a look in how we define server options [here](https://github.com/netdata/netdata/blob/0275a0d468d8b41d4b1d84955b99a77c9eed8759/libnetdata/socket/security.c#L157-L186)

Reply:
how you generate your key and certificate?

Reply:
I was using old TLS certificates that were generated with;

```sh
$ openssl req -new -newkey rsa:2048 -days 365 -nodes -x509 -sha512 -subj "/C=US/ST=Denied/L=Somewhere/O=Dis/CN=www.example.com" -keyout key.pem  -out cert.pem
```